{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'Full_enrichment_ACR_DT'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-e3cbf4a6d056>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mcollections\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mnamedtuple\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 7\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[0mFull_enrichment_ACR_DT\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mipynb\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mProteinSet\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      8\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mread_ppis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'Full_enrichment_ACR_DT'"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scipy.stats as sstats\n",
    "import pandas as pd\n",
    "\n",
    "from collections import namedtuple\n",
    "\n",
    "def read_ppis():\n",
    "    \"\"\"\n",
    "    function: \n",
    "    This function reads in the ppi dataset. \n",
    "    \n",
    "    Variables: \n",
    "    protein_protein = a dataframe with all ppi's that have a higher combined_score than 0.9.\n",
    "    \"\"\"\n",
    "    protein_protein = pd.read_csv('protein_links_v11.0_0.9.tsv', sep=' ')\n",
    "    protein_protein['protein'] = protein_protein['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "    protein_protein['chemical'] = protein_protein['chemical'].map(lambda x: x.lstrip('9606.')) \n",
    "    \n",
    "    return protein_protein\n",
    "\n",
    "def entrez_to_protein(): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function takes a list of entrez gene ID's and turns them into protein ID's. \n",
    "    \n",
    "    Variables: \n",
    "    biomart_data = a dataframe with all the data from biomart. \n",
    "    other_setg = a list of entrez gene ID's that are related to ageing. \n",
    "    get_ensp = a merged datast where biomart_data and other_setg are merged into one dataset by the protein column. \n",
    "    get_ensp_filtered = the same dataset as get_ensp but without some of the columns that are not needed. \n",
    "    \"\"\"\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"protein\", \"Entrez\", \"Uniprot\", \"name\"])\n",
    "    \n",
    "    other_setg = pd.read_csv(\"test_list_genes1.0.txt\")\n",
    "    get_ensp = pd.merge(other_setg, biomart_data, on=[\"Entrez\"]) \n",
    "    get_ensp = get_ensp.dropna(subset=['protein'])\n",
    "    get_ensp_filtered = get_ensp.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    return get_ensp_filtered\n",
    "\n",
    "def make_dictio_ppi(protein_protein, get_ensp_filtered):\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function maps the ppi dataset and the ageing related genes so that it can be enriched later on. \n",
    "    \n",
    "    Variables: \n",
    "    dictio = a dictionary with proteins and a list of proteins that interact with this protein. \n",
    "    filtered_protein = the column with target proteins from the protein_protein dataset filtered by the ageing related genes. \n",
    "    filtered_chemical = the column with the initial protein from the protein_protein dataset filtered by the ageing related \n",
    "    genes. \n",
    "    total = filtered_protein and filtered_chemical put together into one dataset.\n",
    "    \"\"\"\n",
    "    dictio = {}\n",
    "    filtered_protein = protein_protein[protein_protein['protein'].isin(get_ensp_filtered['protein'])]\n",
    "    filtered_chemical = protein_protein[protein_protein['chemical'].isin(get_ensp_filtered['protein'])]\n",
    "    total = pd.concat([filtered_protein, filtered_chemical], ignore_index=True) \n",
    "    \n",
    "    for i in total['chemical'].unique(): \n",
    "        dictio[i] = [total['protein'][j] for j in total[total['chemical']==i].index] \n",
    "    \n",
    "    for x in total['protein'].unique():\n",
    "        if x in dictio:\n",
    "            dictio[x] += ([total['chemical'][y] for y in total[total['protein']==x].index]) \n",
    "        else: \n",
    "            dictio[x] = [total['chemical'][y] for y in total[total['protein']==x].index]\n",
    "    \n",
    "    return dictio \n",
    "    \n",
    "\n",
    "def main_ppi():\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function calls all the important functions and gives them variables. \n",
    "    \n",
    "    Variables: \n",
    "    protein_protein = a dataframe of protein protein interactions returned from the read_ppis() function. \n",
    "    get_ensp_filtered = a dataframe of protein ID's that are related to ageing, returned from the get_ensp_filtered() function. \n",
    "    ppi_dictionary = a dictionary with proteins and lists of proteins that interact with this protein, returned from the \n",
    "    make_dictio_ppi() function. \n",
    "    enrichment_call = calls the class: ProteinSet and gives the class the variable: ppi_dictionary. \n",
    "    df = a dataframe with the results of the enrich() function in the class: ProteinSet. \n",
    "    sort_df = sorts the dataframe on the column pvalue. \n",
    "    \"\"\"\n",
    "    ensembl = pd.read_csv('STITCH_proteins.txt')\n",
    "    protein_protein = read_ppis() \n",
    "    get_ensp_filtered = entrez_to_protein() \n",
    "    ppi_dictionary = make_dictio_ppi(protein_protein, get_ensp_filtered) \n",
    "    enrichment_call = ProteinSet(ppi_dictionary)\n",
    "    df = enrichment_call.enrich(get_ensp_filtered['protein'], ensembl['protein']) \n",
    "    sort_df = df.sort_values('pvalue', ascending=1)\n",
    "    print(sort_df)\n",
    "    \n",
    "\n",
    "main_ppi() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
