{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2.robjects.vectors import StrVector\n",
    "import rpy2.robjects as robjects\n",
    "import rpy2.robjects.packages as rpackages\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as sstats\n",
    "\n",
    "from collections import namedtuple\n",
    "class ProteinSet(object):\n",
    "    def __init__(self, proteindict, database):\n",
    "        \"\"\"\n",
    "        Function: \n",
    "        Making a global variable of the variables that are gonna be used through the whole class. \n",
    "        \n",
    "        Variables: \n",
    "        self.proteindict = a dictionary with a term and a list of proteins per item. \n",
    "        self.database = the name of the database. \n",
    "        \"\"\"\n",
    "        self.proteindict = { name : set(p) for name, p in proteindict.items() }\n",
    "        self.database = database\n",
    "    \n",
    "    def enrich(self, otherset, background):\n",
    "        \"\"\"\n",
    "        Function: \n",
    "        This function takes 3 sets of proteins (or genes) and uses them to make an enrichment using either fisher's exact test or \n",
    "        the chi2 (depending on how big the sets are). \n",
    "        \n",
    "        Variables: \n",
    "        list_res = a list with lists that will later be turned into a dataframe. Eech list within the list will have information\n",
    "        about a row in the table. \n",
    "        name = the name of the drug \n",
    "        pset = a set of proteins that are targets of the drug. \n",
    "        term = the name in a list. \n",
    "        proteins = the proteins in a list. \n",
    "        results = the enrichment results in a NamedTuple. \n",
    "        l_results = the enrichment results turned into a list. \n",
    "        joined = the name of the database and the term merged with the l_results list. \n",
    "        df_final = a dataframe with all the enrichment results. \n",
    "        \"\"\"\n",
    "        list_res = []\n",
    "        for name, pset in self.proteindict.items():\n",
    "            term = [self.database] + [name] \n",
    "            proteins = list(pset)\n",
    "            results = self.set_enrichment(pset, otherset, background)\n",
    "            l_results = list(results)\n",
    "            joined = term + l_results\n",
    "            joined.append(proteins)\n",
    "            list_res.append(joined)\n",
    "        \n",
    "        df_final = pd.DataFrame(list_res)\n",
    "        df_final.columns = ['Database', 'Name', 'oddsratio', 'c2statistic', 'pvalue', 'table', 'method', 'proteins'] \n",
    "        return df_final\n",
    "    \n",
    "    def set_enrichment(self, your_set, other_set, universe, abcd_values=False):\n",
    "    \n",
    "        resTuple = namedtuple(\"setEnrichmentResult\", [ 'oddsratio', 'c2statistic', 'pvalue', 'table', 'method'])\n",
    "\n",
    "        universe  = set(universe)\n",
    "        your_set  = set(your_set) & universe\n",
    "        other_set = set(other_set) & universe\n",
    "\n",
    "        a = your_set & other_set\n",
    "        b = other_set - your_set\n",
    "        c = your_set - other_set\n",
    "        d = universe - (your_set | other_set)\n",
    "\n",
    "        table = [ [len(a), len(b)], [len(c), len(d)]]\n",
    "        if min(min(table)) <= 5:\n",
    "            method = 'fisher'\n",
    "            oddsratio, p = sstats.fisher_exact(table)\n",
    "            chi2 = None\n",
    "        else:\n",
    "            method = 'chi2'\n",
    "            chi2, p, dof, expected = sstats.chi2_contingency(table)\n",
    "            oddsratio = 100\n",
    "            if table[1][0] > 0 and table[0][1] > 0:\n",
    "                oddsratio = table[0][0] * table[1][1] / (table[1][0] * table[0][1])\n",
    "            else:\n",
    "                oddsratio = np.inf\n",
    "\n",
    "        if abcd_values:\n",
    "            return resTuple(oddsratio, chi2, p, [[a,b],[c,d]], method)\n",
    "        else:\n",
    "            return resTuple(oddsratio, chi2, p, table, method)\n",
    "    \n",
    "def protein_to_entrez(other_set1): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function takes a dataset containing ensembl protein id's and turns them into entrez gene id's. \n",
    "    \n",
    "    Variables: \n",
    "    biomart_data = a dataset with entrez gene id's and their corresponding ensembl protein id's. \n",
    "    other_set1 = a list of entrez gene id's. \n",
    "    get_ens = merged dataset with other_set1 and biomart_data. \n",
    "    get_ens_filtered = the same dataset as get_ens but without some of the columns that are not important. \n",
    "    \"\"\"\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"protein\", \"Entrez\", \"Uniprot\", \"name\"])\n",
    "\n",
    "    get_ens = pd.merge(other_set1, biomart_data, on=[\"protein\"]) \n",
    "    get_ens = get_ens.dropna(subset=['Entrez'])\n",
    "    get_ens['Entrez'] = get_ens['Entrez'].astype(int)\n",
    "    get_ens_filtered = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    return get_ens_filtered\n",
    "\n",
    "\n",
    "def entrez_to_protein(other_set1): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function takes a dataset containing Entrez gene id's and turns them into ensembl protein id's. \n",
    "    \n",
    "    Variables: \n",
    "    biomart_data = a dataset with entrez gene id's and their corresponding ensembl protein id's. \n",
    "    other_set1 = a list of entrez gene id's. \n",
    "    get_ens = merged dataset with other_set1 and biomart_data. \n",
    "    get_ens_filtered = the same dataset as get_ens but without some of the columns that are not important. \n",
    "    \"\"\"\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"protein\", \"Entrez\", \"Uniprot\", \"name\"])\n",
    "\n",
    "    get_ens = pd.merge(other_set1, biomart_data, on=[\"Entrez\"]) \n",
    "    get_ens = get_ens.dropna(subset=['protein'])\n",
    "    get_ens_filtered = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    return get_ens_filtered\n",
    "\n",
    "def make_dictio_DT(): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function makes a dictionary with a drug and a list of proteins that are targets of that drug.\n",
    "    \n",
    "    Variable: \n",
    "    mapped = dataset with drugs and their targets. \n",
    "    dictio = a dictionary with a drug and the corresponding targets (proteins).\n",
    "    \"\"\"\n",
    "    mapped = pd.read_csv('mapped_DB_STITCH.tsv', sep='\\t')\n",
    "    mapped['protein'] = mapped['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "    mapped = mapped[['CID', 'InChIKey', 'DrugBank ID', 'Name', 'protein', 'combined_score']].drop_duplicates()\n",
    "    \n",
    "    get_entr_filtered = protein_to_entrez(mapped)\n",
    "    \n",
    "    dictio = {}\n",
    "    for i in get_entr_filtered['Name'].unique(): \n",
    "        dictio[i] = [get_entr_filtered['Entrez'][j] for j in get_entr_filtered[get_entr_filtered['Name']==i].index]\n",
    "    \n",
    "    return dictio \n",
    "    \n",
    "def cluster_profiler_KEGG(clusterProfiler, data): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function uses the R-package: clusterProfiler to enrich a list of genes for KEGG pathways. You will get back a list\n",
    "    of KEGG pathways and the sets of genes part of that pathway. The list is sorted by the adjusted p-value.\n",
    "    \n",
    "    Variables: \n",
    "    enrich_KEGG = the enrichment for the KEGG pathways. This variable is in RS4 format. \n",
    "    KEGGdat = the results of the enrichment in a normal (r) dataframe format.\n",
    "    df = to get the KEGG enrichment results into a pandas dataframe format.\n",
    "    \"\"\"\n",
    "    \n",
    "    enrich_KEGG = clusterProfiler.enrichKEGG(data, organism = 'hsa', keyType = 'kegg', pvalueCutoff = 0.05, pAdjustMethod = 'BY')\n",
    "    KEGGdat = enrich_KEGG.slots['result']\n",
    "    \n",
    "    df = pd.DataFrame(index=range(len(KEGGdat[0])))\n",
    "    df['ID'] = KEGGdat[0]\n",
    "    df['Description'] = KEGGdat[1]\n",
    "    df['GeneRatio'] = KEGGdat[2]\n",
    "    df['pvalue'] = KEGGdat[4]\n",
    "    df['padjust'] = KEGGdat[5]\n",
    "    df['Count'] = KEGGdat[8]\n",
    "    empty_lijst_KEGG = []\n",
    "    for x in KEGGdat[7]:\n",
    "        per_pathway = []\n",
    "        lijst = x.split('/')\n",
    "        lijst = list(map(int, lijst))\n",
    "        empty_lijst_KEGG.append(lijst)\n",
    "    df['GeneID'] = empty_lijst_KEGG\n",
    "    df = df[df['padjust']<0.05]\n",
    "    \n",
    "    dictionary = {} \n",
    "    for index, row in df.iterrows():\n",
    "        dictionary[row['ID']] = row['GeneID']\n",
    "        \n",
    "    return dictionary\n",
    "\n",
    "def cluster_profiler_GO_MF(clusterProfiler, data): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function uses the R-package: clusterProfiler to enrich a list of genes for GO annotations (specifically molecular\n",
    "    function). You will get back a list of GO annotations with the sets of genes that have something to do with that GO anno-\n",
    "    tation. This list is sorted by the adjusted p-value. \n",
    "    \n",
    "    Variables: \n",
    "    enrich_GO_MF = the enrichment for the GO annotations (molecular function). This variable is in RS4 format. \n",
    "    MFdat = the results of the enrichment in a normal (r) dataframe format. \n",
    "    df_GO_MF = to get the GO annotation enrichment results into a pandas dataframe format. \n",
    "    \"\"\"\n",
    "    \n",
    "    enrich_GO_MF = clusterProfiler.enrichGO(data, 'org.Hs.eg.db', ont = 'MF', pvalueCutoff = 0.05, pAdjustMethod = 'BY')\n",
    "    MFdat = enrich_GO_MF.slots['result']\n",
    "    \n",
    "    df_GO_MF = pd.DataFrame(index=range(len(MFdat[0])))\n",
    "    df_GO_MF['ID'] = MFdat[0]\n",
    "    df_GO_MF['Description'] = MFdat[1]\n",
    "    df_GO_MF['GeneRatio'] = MFdat[2]\n",
    "    df_GO_MF['pvalue'] = MFdat[4]\n",
    "    df_GO_MF['padjust'] = MFdat[5]\n",
    "    df_GO_MF['Count'] = MFdat[8]\n",
    "    empty_lijst_GO_MF = []\n",
    "    for x in MFdat[7]: \n",
    "        lijst = x.split('/')\n",
    "        lijst = list(map(int, lijst))\n",
    "        empty_lijst_GO_MF.append(lijst)\n",
    "    df_GO_MF['GeneID'] = empty_lijst_GO_MF\n",
    "    df_GO_MF = df_GO_MF[df_GO_MF['padjust']<0.05]\n",
    "    \n",
    "    dictionary = {} \n",
    "    for index, row in df_GO_MF.iterrows():\n",
    "        dictionary[row['ID']] = row['GeneID']\n",
    "        \n",
    "    return dictionary\n",
    "\n",
    "def cluster_profiler_GO_CC(clusterProfiler, data): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function uses the R-package: clusterProfiler to enrich a list of genes for GO annotations (specifically cellular com-\n",
    "    ponent). You will get back a list of GO annotations with the sets of genes that have something to do with that GO anno-\n",
    "    tation. This list is sorted by the adjusted p-value. \n",
    "    \n",
    "    Variables: \n",
    "    enrich_GO_CC = the enrichment for the GO annotations (cellular component). This variable is in RS4 format. \n",
    "    CCdat = the results of the enrichment in a normal (r) dataframe format. \n",
    "    df_GO_CC = to get the GO annotation enrichment results into a pandas dataframe format.\n",
    "    \"\"\"\n",
    "    \n",
    "    enrich_GO_CC = clusterProfiler.enrichGO(data, 'org.Hs.eg.db', ont = 'CC', pvalueCutoff = 0.05, pAdjustMethod = 'BY')\n",
    "    CCdat = enrich_GO_CC.slots['result']\n",
    "    \n",
    "    df_GO_CC = pd.DataFrame(index=range(len(CCdat[0])))\n",
    "    df_GO_CC['ID'] = CCdat[0]\n",
    "    df_GO_CC['Description'] = CCdat[1]\n",
    "    df_GO_CC['GeneRatio'] = CCdat[2]\n",
    "    df_GO_CC['pvalue'] = CCdat[4]\n",
    "    df_GO_CC['padjust'] = CCdat[5]\n",
    "    df_GO_CC['Count'] = CCdat[8]\n",
    "    empty_lijst_GO_CC = []\n",
    "    for x in CCdat[7]: \n",
    "        lijst = x.split('/')\n",
    "        lijst = list(map(int, lijst))\n",
    "        empty_lijst_GO_CC.append(lijst)\n",
    "    df_GO_CC['GeneID'] = empty_lijst_GO_CC\n",
    "    df_GO_CC = df_GO_CC[df_GO_CC['padjust']<0.05]\n",
    "    \n",
    "    dictionary = {} \n",
    "    for index, row in df_GO_CC.iterrows():\n",
    "        dictionary[row['ID']] = row['GeneID']\n",
    "        \n",
    "    return dictionary\n",
    "\n",
    "def cluster_profiler_GO_BP(clusterProfiler, data):\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function uses the R-package: clusterProfiler to enrich a list of genes for GO annotations (specifically biological\n",
    "    process). You will get back a list of GO annotations with the sets of genes that have something to do with that GO anno-\n",
    "    tation. This list is sorted by the adjusted p-value.\n",
    "    \n",
    "    Variables: \n",
    "    enrich_GO_BP = the enrichment for the GO annotations (biological process). This variable is in RS4 format. \n",
    "    BPdat = the results of the enrichment in a normal (r) dataframe format.\n",
    "    df_GO_BP = to get the GO annotation enrichment results into a pandas dataframe format.\n",
    "    \"\"\"\n",
    "    \n",
    "    enrich_GO_BP = clusterProfiler.enrichGO(data, 'org.Hs.eg.db', ont = 'BP', pvalueCutoff = 0.05, pAdjustMethod = 'BY')\n",
    "    BPdat = enrich_GO_BP.slots['result']\n",
    "    \n",
    "    df_GO_BP = pd.DataFrame(index=range(len(BPdat[0])))\n",
    "    df_GO_BP['ID'] = BPdat[0]\n",
    "    df_GO_BP['Description'] = BPdat[1]\n",
    "    df_GO_BP['GeneRatio'] = BPdat[2]\n",
    "    df_GO_BP['pvalue'] = BPdat[4]\n",
    "    df_GO_BP['padjust'] = BPdat[5]\n",
    "    df_GO_BP['Count'] = BPdat[8]\n",
    "    \n",
    "    empty_lijst_GO_BP = []\n",
    "    for x in BPdat[7]: \n",
    "        lijst = x.split('/')\n",
    "        lijst = list(map(int, lijst))\n",
    "        empty_lijst_GO_BP.append(lijst)\n",
    "        \n",
    "    df_GO_BP['GeneID'] = empty_lijst_GO_BP\n",
    "    df_GO_BP = df_GO_BP[df_GO_BP['padjust']<0.05] \n",
    "    \n",
    "    dictionary = {}\n",
    "    for index, row in df_GO_BP.iterrows():\n",
    "        dictionary[row['ID']] = row['GeneID']\n",
    "    \n",
    "    return dictionary\n",
    "\n",
    "def Reactome(clusterProfiler, ReactomePA, data): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function uses the R-package: ReactomePA to enrich a list of genes for Reactome pathways. You will get back a list of \n",
    "    Reactome pathways with the sets of genes that have something to do with that Reactome pathway. This list will be sorted \n",
    "    by the adjusted p-value. \n",
    "    \n",
    "    Variable: \n",
    "    enrich_Reactome = the enrichment for the reactome pathways. This variable is in RS4 format. \n",
    "    ReactomeDat = the results of the enrichment in a normal (r) dataframe format. \n",
    "    df_Reactome = to get the Reactome enrichment results into a pandas dataframe format. \n",
    "    \"\"\"\n",
    "    \n",
    "    enrich_Reactome = ReactomePA.enrichPathway(gene=data, pvalueCutoff = 0.05, readable = True, pAdjustMethod = 'BY', organism = \"human\")\n",
    "    ReactomeDat = enrich_Reactome.slots['result']\n",
    "    \n",
    "    df_Reactome = pd.DataFrame(index=range(len(ReactomeDat[0])))\n",
    "    df_Reactome['ID'] = ReactomeDat[0]\n",
    "    df_Reactome['Description'] = ReactomeDat[1]\n",
    "    df_Reactome['GeneRatio'] = ReactomeDat[2]\n",
    "    df_Reactome['pvalue'] = ReactomeDat[4]\n",
    "    df_Reactome['padjust'] = ReactomeDat[5]\n",
    "    df_Reactome['Count'] = ReactomeDat[8]\n",
    "    \n",
    "    entrez_list = []\n",
    "    for x in ReactomeDat[7]:\n",
    "        input_bitr = x.split('/')\n",
    "        eg = clusterProfiler.bitr(input_bitr, fromType=\"SYMBOL\", toType=\"ENTREZID\", OrgDb=\"org.Hs.eg.db\")\n",
    "        list_gene_string = list(eg[1])\n",
    "        list_gene_int = list(map(int, list_gene_string))\n",
    "        entrez_list.append(list_gene_int)\n",
    "        \n",
    "\n",
    "    df_Reactome['GeneID'] = entrez_list\n",
    "    df_Reactome = df_Reactome[df_Reactome['padjust']<0.05]\n",
    "    \n",
    "    dictionary = {} \n",
    "    for index, row in df_Reactome.iterrows():\n",
    "        dictionary[row['ID']] = row['GeneID']\n",
    "    \n",
    "    return dictionary\n",
    "\n",
    "def ppi_interactions(gene_list):\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function gets the protein-protein interactions in the  from the r-package: STRINGdb. \n",
    "    \"\"\"\n",
    "    \n",
    "    STRINGdb = rpackages.importr('STRINGdb')\n",
    "    robjects.r(\"string_db <- STRINGdb$new(score_threshold=900)\")\n",
    "    robjects.r(\"df <- data.frame(Entrez=matrix(unlist(%s), nrow=length(%s), byrow=T))\" % (gene_list.r_repr(), gene_list.r_repr()))  \n",
    "    robjects.r('mapped <- string_db$map(df, \"Entrez\", removeUnmappedRows = TRUE)')\n",
    "    interactions = robjects.r(\"string_db$get_interactions(mapped$STRING_id)\") \n",
    "    \n",
    "    return interactions\n",
    "    \n",
    "def ppi_dictio(gene_list): \n",
    "    \n",
    "    interactions = ppi_interactions(gene_list)\n",
    "    chemical = list(interactions[0])\n",
    "    df_proteins = pd.DataFrame(list(interactions[1]))\n",
    "    df_proteins.columns = ['protein']\n",
    "    df_proteins['chemical'] = chemical\n",
    "    df_proteins['protein'] = df_proteins['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "    df_proteins['chemical'] = df_proteins['chemical'].map(lambda x: x.lstrip('9606.'))\n",
    "\n",
    "    dictio = {}\n",
    "\n",
    "    protein_gene = protein_to_entrez(df_proteins)\n",
    "    protein_gene = protein_gene.rename(columns={\"Entrez\": \"Entrez_protein\"})\n",
    "\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                    sep='\\t', \n",
    "                    names=[\"gene\", \"transcript\", \"chemical\", \"Entrez_chemical\", \"Uniprot\", \"name\"])\n",
    "\n",
    "\n",
    "    get_ens = pd.merge(protein_gene, biomart_data, on=[\"chemical\"]) \n",
    "    get_ens = get_ens.dropna(subset=['Entrez_chemical'])\n",
    "    get_ens['Entrez_chemical'] = get_ens['Entrez_chemical'].astype(int)\n",
    "    chem_prot_gene = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "\n",
    "    for i in chem_prot_gene['Entrez_chemical'].unique(): \n",
    "        dictio[i] = [chem_prot_gene['Entrez_protein'][j] for j in chem_prot_gene[chem_prot_gene['Entrez_chemical']==i].index] \n",
    "\n",
    "    for x in chem_prot_gene['Entrez_protein'].unique():\n",
    "        if x in dictio:\n",
    "            dictio[x] += ([chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]) \n",
    "        else: \n",
    "            dictio[x] = [chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]\n",
    "\n",
    "    for pro, chem in dictio.items(): \n",
    "        dictio[pro] = list(set(chem))\n",
    "\n",
    "    return dictio \n",
    "\n",
    "\n",
    "#def read_ppis():\n",
    "#    \"\"\"\n",
    "#    function: \n",
    "#    This function reads in the ppi dataset. \n",
    "#    \n",
    "#    Variables: \n",
    "#    protein_protein = a dataframe with all ppi's that have a higher combined_score than 0.9.\n",
    "#    \"\"\"\n",
    "#    protein_protein = pd.read_csv('protein_links_v11.0_0.9.tsv', sep=' ')\n",
    "#    protein_protein['protein'] = protein_protein['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "#    protein_protein['chemical'] = protein_protein['chemical'].map(lambda x: x.lstrip('9606.'))  \n",
    "#    \n",
    "#    return protein_protein\n",
    "\n",
    "#def make_dictio_ppi(protein_protein, get_ensp_filtered):\n",
    "#    \"\"\"\n",
    "#    Function: \n",
    "#    This function maps the ppi dataset and the ageing related genes so that it can be enriched later on. \n",
    "#    \n",
    "#    Variables: \n",
    "#    dictio = a dictionary with proteins and a list of proteins that interact with this protein. \n",
    "#    filtered_protein = the column with target proteins from the protein_protein dataset filtered by the ageing related genes. \n",
    "#    filtered_chemical = the column with the initial protein from the protein_protein dataset filtered by the ageing related \n",
    "#    genes. \n",
    "#    total = filtered_protein and filtered_chemical put together into one dataset.\n",
    "#    protein_gene = converts the protein column of the total dataframe to gene Entrez ID's. \n",
    "#    biomart_data = a dataframe with all the data from biomart. \n",
    "#    get_ens = merges the biomart dataset and the protein_gene dataset to get the Entrez ID's for the chemicals. \n",
    "#    chem_prot_gene = a dataframe with Entrez ID's for the proteins. \n",
    "#    dictio = a dictionary with genes and a list of genes that that gene interacts with. \n",
    "#    \"\"\"\n",
    "#    dictio = {}\n",
    "#    filtered_protein = protein_protein[protein_protein['protein'].isin(get_ensp_filtered['protein'])]\n",
    "#    filtered_chemical = protein_protein[protein_protein['chemical'].isin(get_ensp_filtered['protein'])]\n",
    "#    total = pd.concat([filtered_protein, filtered_chemical], ignore_index=True) \n",
    "#    \n",
    "#    protein_gene = protein_to_entrez(total)\n",
    "#    protein_gene = protein_gene.rename(columns={\"Entrez\": \"Entrez_protein\"})\n",
    "#    \n",
    "#    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "#                  sep='\\t', \n",
    "#                  names=[\"gene\", \"transcript\", \"chemical\", \"Entrez_chemical\", \"Uniprot\", \"name\"])\n",
    "#\n",
    "#    get_ens = pd.merge(protein_gene, biomart_data, on=[\"chemical\"]) \n",
    "#    get_ens = get_ens.dropna(subset=['Entrez_chemical'])\n",
    "#    get_ens['Entrez_chemical'] = get_ens['Entrez_chemical'].astype(int)\n",
    "#    chem_prot_gene = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "#    \n",
    "#    for i in chem_prot_gene['Entrez_chemical'].unique(): \n",
    "#        dictio[i] = [chem_prot_gene['Entrez_protein'][j] for j in chem_prot_gene[chem_prot_gene['Entrez_chemical']==i].index] \n",
    "#    \n",
    "#    for x in chem_prot_gene['Entrez_protein'].unique():\n",
    "#        if x in dictio:\n",
    "#            dictio[x] += ([chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]) \n",
    "#        else: \n",
    "#            dictio[x] = [chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]\n",
    "#    \n",
    "#    for pro, chem in dictio.items(): \n",
    "#        dictio[pro] = list(set(chem))\n",
    "#        \n",
    "#    return dictio \n",
    "\n",
    "\n",
    "class main_enrichments(object):\n",
    "    def __init__(self, gene_list):\n",
    "        \"\"\"\n",
    "        Function: \n",
    "        This function calls all the other functions that have something to do with enriching with R. It also imports the needed\n",
    "        R-packages. Before this function can be called please make sure you have the packages already installed and the packages \n",
    "        are in the right folder (/Users/user/Anaconda3/Lib/R/library). \n",
    "\n",
    "        Variables: \n",
    "        self.ensembl = all the target proteins from the STITCH database. \n",
    "        self.get_entr_filtered_ens = all the ensembl proteins turned into gene ID's.\n",
    "        KEGG_res = results of the first enrichment with the KEGG database in a dictionary. \n",
    "        GO_MF_en_res = results of the first enrichment with the GO molecular functions database in a dictionary.\n",
    "        GO_CC_en_res = results of the first enrichment with the GO cellular component database in a dictionary. \n",
    "        GO_BP_en_res = results of the first enrichment with the GO biological process database in a dictionary. \n",
    "        reactome_en_res = results of the first enrichment with the reactome database in a dictionary. \n",
    "        acr_res = dictionary of gene_list: and a list of genes. \n",
    "        gene_set = a list of ageing related genes.\n",
    "        clusterProfiler = the package: clusterProfiler imported using rpy2.\n",
    "        ReactomePA = the package: ReactomePA imported using rpy2.\n",
    "        dictio = dictionary of drugs and lists of targets. \n",
    "        databases = list of all the results from the first enrichment. \n",
    "        db_names = list containing all the names of databases. \n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        ensembl = pd.read_csv('STITCH_proteins.txt')\n",
    "        self.ensembl = ensembl\n",
    "\n",
    "        get_entr_filtered_ens = protein_to_entrez(ensembl)\n",
    "        self.get_entr_filtered_ens = get_entr_filtered_ens\n",
    "        \n",
    "        gene_set = list(gene_list)\n",
    "        gene_set = list(map(int, gene_set))\n",
    "        \n",
    "        #genes = pd.DataFrame(gene_set)\n",
    "        #genes.columns = ['Entrez']\n",
    "        #genes['Entrez']= genes['Entrez'].astype(float)\n",
    "        #get_ensp_filtered = entrez_to_protein(genes)\n",
    "        #protein_protein = read_ppis()\n",
    "\n",
    "        clusterProfiler = rpackages.importr('clusterProfiler')\n",
    "        ReactomePA = rpackages.importr('ReactomePA')\n",
    "\n",
    "        print(\"prep kegg\")\n",
    "        KEGG_res = cluster_profiler_KEGG(clusterProfiler, gene_list)\n",
    "        print(\"prep go_mf\")\n",
    "        GO_MF_en_res = cluster_profiler_GO_MF(clusterProfiler, gene_list)\n",
    "        print(\"prep go_cc\")\n",
    "        GO_CC_en_res = cluster_profiler_GO_CC(clusterProfiler, gene_list)\n",
    "        print(\"prep go_bp\")\n",
    "        GO_BP_en_res = cluster_profiler_GO_BP(clusterProfiler, gene_list)\n",
    "        print(\"prep reactome\")\n",
    "        reactome_en_res = Reactome(clusterProfiler, ReactomePA, gene_list)\n",
    "        print(\"prep gene list\")\n",
    "        acr_res = { \"gene_list\" : gene_set }\n",
    "        print(\"prep PPI's\")\n",
    "        ppi_res = ppi_dictio(gene_list) \n",
    "        #ppi_res = make_dictio_ppi(protein_protein, get_ensp_filtered)\n",
    "        \n",
    "\n",
    "        dictio = make_dictio_DT()\n",
    "        databases = [KEGG_res, GO_MF_en_res, GO_CC_en_res, GO_BP_en_res, reactome_en_res, acr_res, ppi_res]\n",
    "        db_names  = ['KEGG', 'GO_MF', 'GO_CC', 'GO_BP', 'Reactome', 'gene_list', 'STRING']\n",
    "\n",
    "        self.dictio_drugs = make_dictio_DT()\n",
    "        #self.dictio_drugs = dict(list(self.dictio_drugs.items())[:20])\n",
    "\n",
    "        self.PS = [ProteinSet(db,name) for (db,name) in zip(databases, db_names)]\n",
    "    \n",
    "    def enrich_BL(self):\n",
    "\n",
    "        super_x = []\n",
    "        for i, (drugs, targets) in enumerate(self.dictio_drugs.items()):\n",
    "            print(\"%d/%d: %s\" % (i+1, len(self.dictio_drugs), drugs))\n",
    "            df = pd.concat([ps.enrich(targets, self.get_entr_filtered_ens['Entrez']) for ps in self.PS])\n",
    "            df['drug'] = drugs\n",
    "            df = df.sort_values(\"pvalue\").groupby(\"Database\", as_index=False).first() \n",
    "            super_x.append(df)\n",
    "        self.drug_enrichments = pd.concat(super_x)\n",
    "        return self.drug_enrichments\n",
    "    \n",
    "\n",
    "ACR_genes = robjects.r(\"scan('/home/mhaan/test_list_genes1.0.txt', what='', sep='\\n', skip = 1)\")\n",
    "ME = main_enrichments(ACR_genes)\n",
    "enrich_ME = ME.enrich_BL() \n",
    "print(enrich_ME)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
