{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Database    Name  oddsratio c2statistic  pvalue                 table  \\\n",
      "0      STRING     381        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "1      STRING    3835        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "2      STRING    9183        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "3      STRING     662        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "4      STRING    1315        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "...       ...     ...        ...         ...     ...                   ...   \n",
      "8219   STRING  137872        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8220   STRING  164668        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8221   STRING  256987        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8222   STRING    6876        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8223   STRING    1837        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "\n",
      "      method                                           proteins      drug  \n",
      "0     fisher  [64837, 23095, 23303, 1639, 3630, 3798, 6711, ...  CTT-1057  \n",
      "1     fisher                  [64837, 1639, 23303, 3798, 23095]  CTT-1057  \n",
      "2     fisher             [5347, 64837, 23303, 3798, 23095, 701]  CTT-1057  \n",
      "3     fisher                        [23095, 64837, 3798, 23303]  CTT-1057  \n",
      "4     fisher  [64837, 23095, 23303, 1639, 3630, 2548, 79895,...  CTT-1057  \n",
      "...      ...                                                ...       ...  \n",
      "8219  fisher                                              [128]  CTT-1057  \n",
      "8220  fisher  [100913187, 9582, 403314, 10930, 339, 27350, 2...  CTT-1057  \n",
      "8221  fisher                                            [10227]  CTT-1057  \n",
      "8222  fisher                                            [93649]  CTT-1057  \n",
      "8223  fisher                                      [54221, 6645]  CTT-1057  \n",
      "\n",
      "[8224 rows x 9 columns]\n",
      "     Database    Name  oddsratio c2statistic  pvalue                 table  \\\n",
      "0      STRING     381        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "1      STRING    3835        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "2      STRING    9183        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "3      STRING     662        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "4      STRING    1315        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "...       ...     ...        ...         ...     ...                   ...   \n",
      "8219   STRING  137872        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8220   STRING  164668        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8221   STRING  256987        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8222   STRING    6876        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8223   STRING    1837        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "\n",
      "      method                                           proteins  \\\n",
      "0     fisher  [64837, 23095, 23303, 1639, 3630, 3798, 6711, ...   \n",
      "1     fisher                  [64837, 1639, 23303, 3798, 23095]   \n",
      "2     fisher             [5347, 64837, 23303, 3798, 23095, 701]   \n",
      "3     fisher                        [23095, 64837, 3798, 23303]   \n",
      "4     fisher  [64837, 23095, 23303, 1639, 3630, 2548, 79895,...   \n",
      "...      ...                                                ...   \n",
      "8219  fisher                                              [128]   \n",
      "8220  fisher  [100913187, 9582, 403314, 10930, 339, 27350, 2...   \n",
      "8221  fisher                                            [10227]   \n",
      "8222  fisher                                            [93649]   \n",
      "8223  fisher                                      [54221, 6645]   \n",
      "\n",
      "                                                   drug  \n",
      "0     (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "1     (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "2     (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "3     (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "4     (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "...                                                 ...  \n",
      "8219  (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "8220  (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "8221  (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "8222  (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "8223  (2S)-2-{[HYDROXY(4-IODOBENZYL)PHOSPHORYL]METHY...  \n",
      "\n",
      "[8224 rows x 9 columns]\n",
      "     Database    Name  oddsratio c2statistic  pvalue                 table  \\\n",
      "0      STRING     381        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "1      STRING    3835        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "2      STRING    9183        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "3      STRING     662        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "4      STRING    1315        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "...       ...     ...        ...         ...     ...                   ...   \n",
      "8219   STRING  137872        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8220   STRING  164668        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8221   STRING  256987        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8222   STRING    6876        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "8223   STRING    1837        NaN        None     1.0  [[0, 0], [0, 19195]]   \n",
      "\n",
      "      method                                           proteins  \\\n",
      "0     fisher  [64837, 23095, 23303, 1639, 3630, 3798, 6711, ...   \n",
      "1     fisher                  [64837, 1639, 23303, 3798, 23095]   \n",
      "2     fisher             [5347, 64837, 23303, 3798, 23095, 701]   \n",
      "3     fisher                        [23095, 64837, 3798, 23303]   \n",
      "4     fisher  [64837, 23095, 23303, 1639, 3630, 2548, 79895,...   \n",
      "...      ...                                                ...   \n",
      "8219  fisher                                              [128]   \n",
      "8220  fisher  [100913187, 9582, 403314, 10930, 339, 27350, 2...   \n",
      "8221  fisher                                            [10227]   \n",
      "8222  fisher                                            [93649]   \n",
      "8223  fisher                                      [54221, 6645]   \n",
      "\n",
      "                  drug  \n",
      "0     O-Sulfo-L-Serine  \n",
      "1     O-Sulfo-L-Serine  \n",
      "2     O-Sulfo-L-Serine  \n",
      "3     O-Sulfo-L-Serine  \n",
      "4     O-Sulfo-L-Serine  \n",
      "...                ...  \n",
      "8219  O-Sulfo-L-Serine  \n",
      "8220  O-Sulfo-L-Serine  \n",
      "8221  O-Sulfo-L-Serine  \n",
      "8222  O-Sulfo-L-Serine  \n",
      "8223  O-Sulfo-L-Serine  \n",
      "\n",
      "[8224 rows x 9 columns]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-876e92af8252>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    263\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfObj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 265\u001b[0;31m \u001b[0mmain_ppi\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    266\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    267\u001b[0m \u001b[0mmain_ACR_DT\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-17-876e92af8252>\u001b[0m in \u001b[0;36mmain_ppi\u001b[0;34m()\u001b[0m\n\u001b[1;32m    257\u001b[0m     \u001b[0mdfObj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    258\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mdrugs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtargets\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdictio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 259\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0menrichment_call\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menrich\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtargets\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensembl\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'protein'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    260\u001b[0m         \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'drug'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdrugs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-17-876e92af8252>\u001b[0m in \u001b[0;36menrich\u001b[0;34m(self, otherset, background)\u001b[0m\n\u001b[1;32m     40\u001b[0m             \u001b[0mterm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatabase\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m             \u001b[0mproteins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m             \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_enrichment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0motherset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbackground\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     43\u001b[0m             \u001b[0ml_results\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m             \u001b[0mjoined\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mterm\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0ml_results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scipy.stats as sstats\n",
    "import pandas as pd \n",
    "\n",
    "from collections import namedtuple\n",
    "\n",
    "class ProteinSet(object):\n",
    "    def __init__(self, proteindict, database):\n",
    "        \"\"\"\n",
    "        Function: \n",
    "        Making a global variable of the variables that are gonna be used through the whole class. \n",
    "        \n",
    "        Variables: \n",
    "        self.proteindict = a dictionary with a term and a list of proteins per item. \n",
    "        self.database = the name of the database. \n",
    "        \"\"\"\n",
    "        self.proteindict = { name : set(p) for name, p in proteindict.items() }\n",
    "        self.database = database\n",
    "    \n",
    "    def enrich(self, otherset, background):\n",
    "        \"\"\"\n",
    "        Function: \n",
    "        This function takes 3 sets of proteins (or genes) and uses them to make an enrichment using either fisher's exact test or \n",
    "        the chi2 (depending on how big the sets are). \n",
    "        \n",
    "        Variables: \n",
    "        list_res = a list with lists that will later be turned into a dataframe. Eech list within the list will have information\n",
    "        about a row in the table. \n",
    "        name = the name of the drug \n",
    "        pset = a set of proteins that are targets of the drug. \n",
    "        term = the name in a list. \n",
    "        proteins = the proteins in a list. \n",
    "        results = the enrichment results in a NamedTuple. \n",
    "        l_results = the enrichment results turned into a list. \n",
    "        joined = the name of the database and the term merged with the l_results list. \n",
    "        df_final = a dataframe with all the enrichment results. \n",
    "        \"\"\"\n",
    "        list_res = []\n",
    "        for name, pset in self.proteindict.items():\n",
    "            term = [self.database] + [name] \n",
    "            proteins = list(pset)\n",
    "            results = self.set_enrichment(pset, otherset, background)\n",
    "            l_results = list(results)\n",
    "            joined = term + l_results\n",
    "            joined.append(proteins)\n",
    "            list_res.append(joined)\n",
    "        \n",
    "        df_final = pd.DataFrame(list_res)\n",
    "        df_final.columns = ['Database', 'Name', 'oddsratio', 'c2statistic', 'pvalue', 'table', 'method', 'proteins'] \n",
    "        return df_final\n",
    "    \n",
    "    def set_enrichment(self, your_set, other_set, universe, abcd_values=False):\n",
    "    \n",
    "        resTuple = namedtuple(\"setEnrichmentResult\", [ 'oddsratio', 'c2statistic', 'pvalue', 'table', 'method'])\n",
    "\n",
    "        universe  = set(universe)\n",
    "        your_set  = set(your_set) & universe\n",
    "        other_set = set(other_set) & universe\n",
    "\n",
    "        a = your_set & other_set\n",
    "        b = other_set - your_set\n",
    "        c = your_set - other_set\n",
    "        d = universe - (your_set | other_set)\n",
    "\n",
    "        table = [ [len(a), len(b)], [len(c), len(d)]]\n",
    "        if min(min(table)) <= 5:\n",
    "            method = 'fisher'\n",
    "            oddsratio, p = sstats.fisher_exact(table)\n",
    "            chi2 = None\n",
    "        else:\n",
    "            method = 'chi2'\n",
    "            chi2, p, dof, expected = sstats.chi2_contingency(table)\n",
    "            oddsratio = 100\n",
    "            if table[1][0] > 0 and table[0][1] > 0:\n",
    "                oddsratio = table[0][0] * table[1][1] / (table[1][0] * table[0][1])\n",
    "            else:\n",
    "                oddsratio = np.inf\n",
    "\n",
    "        if abcd_values:\n",
    "            return resTuple(oddsratio, chi2, p, [[a,b],[c,d]], method)\n",
    "        else:\n",
    "            return resTuple(oddsratio, chi2, p, table, method)\n",
    "\n",
    "def make_dictio_DT(): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function makes a dictionary with a drug and a list of proteins that are targets of that drug.\n",
    "    \n",
    "    Variable: \n",
    "    mapped = dataset with drugs and their targets. \n",
    "    dictio = a dictionary with a drug and the corresponding targets (proteins).\n",
    "    \"\"\"\n",
    "    mapped = pd.read_csv('mapped_DB_STITCH.tsv', sep='\\t')\n",
    "    mapped['protein'] = mapped['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "    mapped = mapped[['CID', 'InChIKey', 'DrugBank ID', 'Name', 'protein', 'combined_score']].drop_duplicates()\n",
    "    \n",
    "    get_entr_filtered = protein_to_entrez(mapped)\n",
    "    \n",
    "    dictio = {}\n",
    "    for i in get_entr_filtered['Name'].unique(): \n",
    "        dictio[i] = [get_entr_filtered['Entrez'][j] for j in get_entr_filtered[get_entr_filtered['Name']==i].index]\n",
    "    \n",
    "    return dictio \n",
    "\n",
    "def protein_to_entrez(other_set1): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function takes a dataset containing ensembl protein id's and turns them into entrez gene id's. \n",
    "    \n",
    "    Variables: \n",
    "    biomart_data = a dataset with entrez gene id's and their corresponding ensembl protein id's. \n",
    "    other_set1 = a list of entrez gene id's. \n",
    "    get_ens = merged dataset with other_set1 and biomart_data. \n",
    "    get_ens_filtered = the same dataset as get_ens but without some of the columns that are not important. \n",
    "    \"\"\"\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"protein\", \"Entrez\", \"Uniprot\", \"name\"])\n",
    "\n",
    "    get_ens = pd.merge(other_set1, biomart_data, on=[\"protein\"]) \n",
    "    get_ens = get_ens.dropna(subset=['Entrez'])\n",
    "    get_ens['Entrez'] = get_ens['Entrez'].astype(int)\n",
    "    get_ens_filtered = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    return get_ens_filtered\n",
    "\n",
    "\n",
    "def entrez_to_protein(other_set1): \n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function takes a dataset containing ensembl protein id's and turns them into entrez gene id's. \n",
    "    \n",
    "    Variables: \n",
    "    biomart_data = a dataset with entrez gene id's and their corresponding ensembl protein id's. \n",
    "    other_set1 = a list of entrez gene id's. \n",
    "    get_ens = merged dataset with other_set1 and biomart_data. \n",
    "    get_ens_filtered = the same dataset as get_ens but without some of the columns that are not important. \n",
    "    \"\"\"\n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"protein\", \"Entrez\", \"Uniprot\", \"name\"])\n",
    "\n",
    "    get_ens = pd.merge(other_set1, biomart_data, on=[\"Entrez\"]) \n",
    "    get_ens = get_ens.dropna(subset=['protein'])\n",
    "    get_ens_filtered = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    return get_ens_filtered\n",
    "\n",
    "def main_ACR_DT():\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function calls all the functions and reads in the universe for the enrichment. \n",
    "    \n",
    "    Variables: \n",
    "    ensembl = the universe used for the enrichment. \n",
    "    dictio = the dictionary of drugs and lists of corresponding targets. \n",
    "    get_ens_filtered = a dataframe with entrez gene id's and their corresponding protein id's. \n",
    "    enrichment_call = calls the class: ProteinSet(dictio)\n",
    "    df = a dataframe with all the enrichment results. \n",
    "    \"\"\"\n",
    "    ensembl = pd.read_csv('STITCH_proteins.txt')\n",
    "    get_entr_filtered_ens = protein_to_entrez(ensembl)\n",
    "    \n",
    "    gene_set = pd.read_csv(\"test_list_genes1.0.txt\")\n",
    "    \n",
    "    dictio = make_dictio_DT()\n",
    "    database = \"Genes\"\n",
    "    enrichment_call = ProteinSet(dictio, database)\n",
    "    df = enrichment_call.enrich(gene_set['Entrez'], get_entr_filtered_ens['Entrez']) \n",
    "    print(df)\n",
    "    \n",
    "def read_ppis():\n",
    "    \"\"\"\n",
    "    function: \n",
    "    This function reads in the ppi dataset. \n",
    "    \n",
    "    Variables: \n",
    "    protein_protein = a dataframe with all ppi's that have a higher combined_score than 0.9.\n",
    "    \"\"\"\n",
    "    protein_protein = pd.read_csv('protein_links_v11.0_0.9.tsv', sep=' ')\n",
    "    protein_protein['protein'] = protein_protein['protein'].map(lambda x: x.lstrip('9606.'))\n",
    "    protein_protein['chemical'] = protein_protein['chemical'].map(lambda x: x.lstrip('9606.')) \n",
    "    #protein_protein = pd.read_csv('string_interactions.tsv', sep='\\t')\n",
    "    #print(protein_protein)\n",
    "    #protein_protein['node2_external_id'] = protein_protein['node2_external_id'].map(lambda x: x.lstrip('9606.'))\n",
    "    #protein_protein['node1_external_id'] = protein_protein['node1_external_id'].map(lambda x: x.lstrip('9606.')) \n",
    "    \n",
    "    return protein_protein\n",
    "\n",
    "def make_dictio_ppi(protein_protein, get_ensp_filtered):\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function maps the ppi dataset and the ageing related genes so that it can be enriched later on. \n",
    "    \n",
    "    Variables: \n",
    "    dictio = a dictionary with proteins and a list of proteins that interact with this protein. \n",
    "    filtered_protein = the column with target proteins from the protein_protein dataset filtered by the ageing related genes. \n",
    "    filtered_chemical = the column with the initial protein from the protein_protein dataset filtered by the ageing related \n",
    "    genes. \n",
    "    total = filtered_protein and filtered_chemical put together into one dataset.\n",
    "    \"\"\"\n",
    "    dictio = {}\n",
    "    filtered_protein = protein_protein[protein_protein['protein'].isin(get_ensp_filtered['protein'])]\n",
    "    filtered_chemical = protein_protein[protein_protein['chemical'].isin(get_ensp_filtered['protein'])]\n",
    "    total = pd.concat([filtered_protein, filtered_chemical], ignore_index=True) \n",
    "    \n",
    "    protein_gene = protein_to_entrez(total)\n",
    "    protein_gene = protein_gene.rename(columns={\"Entrez\": \"Entrez_protein\"})\n",
    "    \n",
    "    biomart_data = pd.read_csv(\"biomart.tsv\", \n",
    "                  sep='\\t', \n",
    "                  names=[\"gene\", \"transcript\", \"chemical\", \"Entrez_chemical\", \"Uniprot\", \"name\"])\n",
    "\n",
    "    get_ens = pd.merge(protein_gene, biomart_data, on=[\"chemical\"]) \n",
    "    get_ens = get_ens.dropna(subset=['Entrez_chemical'])\n",
    "    get_ens['Entrez_chemical'] = get_ens['Entrez_chemical'].astype(int)\n",
    "    chem_prot_gene = get_ens.drop([\"gene\", \"transcript\", \"Uniprot\"], axis=1)\n",
    "    \n",
    "    for i in chem_prot_gene['Entrez_chemical'].unique(): \n",
    "        dictio[i] = [chem_prot_gene['Entrez_protein'][j] for j in chem_prot_gene[chem_prot_gene['Entrez_chemical']==i].index] \n",
    "    \n",
    "    for x in chem_prot_gene['Entrez_protein'].unique():\n",
    "        if x in dictio:\n",
    "            dictio[x] += ([chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]) \n",
    "        else: \n",
    "            dictio[x] = [chem_prot_gene['Entrez_chemical'][y] for y in chem_prot_gene[chem_prot_gene['Entrez_protein']==x].index]\n",
    "    \n",
    "    return dictio \n",
    "    \n",
    "\n",
    "def main_ppi():\n",
    "    \"\"\"\n",
    "    Function: \n",
    "    This function calls all the important functions and gives them variables. \n",
    "    \n",
    "    Variables: \n",
    "    protein_protein = a dataframe of protein protein interactions returned from the read_ppis() function. \n",
    "    get_ensp_filtered = a dataframe of protein ID's that are related to ageing, returned from the get_ensp_filtered() function. \n",
    "    ppi_dictionary = a dictionary with proteins and lists of proteins that interact with this protein, returned from the \n",
    "    make_dictio_ppi() function. \n",
    "    enrichment_call = calls the class: ProteinSet and gives the class the variable: ppi_dictionary. \n",
    "    df = a dataframe with the results of the enrich() function in the class: ProteinSet. \n",
    "    sort_df = sorts the dataframe on the column pvalue. \n",
    "    \"\"\"\n",
    "    ensembl = pd.read_csv('STITCH_proteins.txt')\n",
    "    get_entr_filtered_ens = protein_to_entrez(ensembl)\n",
    "    \n",
    "    gene_set = pd.read_csv(\"test_list_genes1.0.txt\")\n",
    "    \n",
    "    protein_protein = read_ppis() \n",
    "    \n",
    "    get_ensp_filtered = entrez_to_protein(gene_set)\n",
    "    ppi_dictionary = make_dictio_ppi(protein_protein, get_ensp_filtered)\n",
    "    dictio = make_dictio_DT()\n",
    "    enrichment_call = ProteinSet(ppi_dictionary, ppi)\n",
    "    dfObj = pd.DataFrame()\n",
    "    \n",
    "\n",
    "main_ppi() \n",
    "\n",
    "main_ACR_DT()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
